### ISSUE TITLE
Asynchronous shader reflection information

### ISSUE THREAD

**Issue by litherum** [gpuweb/gpuweb#624]
During the last F2F, the idea of exposing reflection information came up. There are a few different use cases:

1. Something like Shadertoy, where the JS might not know the content of the shaders it's ingesting. (Shadertoy specifically solves this by requiring a particular interface to every shader, but it's not unrealistic to imagine there are places which wouldn't have this requirement.)
2. During the last F2F, the D3D team indicates that many AAA games create giant shaders with tons of unused inputs/outputs, and then ask the shader compiler which inputs/outputs are actually used, so they can hook up only those resources.
3. Could be used to solve the concerns in https://github.com/gpuweb/gpuweb/issues/561

As far as I know, all the native APIs provide some sort of shader reflection information.

(There's no "API" GitHub tag, so I don't know how to tag this issue as both "relevant to WGSL" but also "relevant to the API")

**Comment by kvark** [comment]
Thank you for filing this!

It's not in the spirit of low-level APIs to expose any information that the user is already supposed to know. However, it is established in the graphics community that there *are* applications that work best with this approach. This includes AAA games with large variety of shaders, as well as things like [Granite](https://themaister.net/blog/2019/04/20/a-tour-of-granites-vulkan-backend-part-3/). So I think it's important for us to allow these use cases to exist.

> As far as I know, all the native APIs provide some sort of shader reflection information.

It's not clear to me if this matters. We could extract all the relevant info from the WGSL module ourselves.

**Comment by kainino0x** [comment]
Observation: Any engine that is generating WGSL already should have enough reflection data in userspace. So adding this would only help engines which are WGSL-first.

**Comment by kainino0x** [comment]
(Notably, I think this makes it reasonable to mark post-MVP.)

**Comment by litherum** [comment]
> > As far as I know, all the native APIs provide some sort of shader reflection information.
> 
> It's not clear to me if this matters. We could extract all the relevant info from the WGSL module ourselves.

I included this not as an an implementation concern, but instead as more of a motivation for the issue. The fact that all the native APIs include this functionality is an indication that authors want it.

**Comment by magcius** [comment]
> As far as I know, all the native APIs provide some sort of shader reflection information.

DX12 and Vulkan provide no reflection information, as far as I know. DirectXShaderCompiler provides some, but that's offline. In Vulkan, community tooling like SPIRV-Reflect is often used.

**Comment by litherum** [comment]
Perhaps this could be integrated with https://github.com/gpuweb/gpuweb/pull/646

**Comment by greggman** [comment]
So ............ Of course my experience with the API is still limited but I think this needs more consideration. 

The library I suggested above is currently relatively small but it turns out support for `const` and `alias` is missing. `alias` seems trivial to add but `const` is more problematic. Consider this valid WGSL

```
          const FOO = radians(90);
          const BAR = sin(FOO);
          const NUM_COLORS = u32(FOO + 3); // result 4
          @group(0) @binding(0) var<uniform> uni: array<vec4f, NUM_COLORS>;
```

[Working example](https://jsgist.org/?src=bcebe5c22fc9ed9ecc18f46fdf986ffc)

Sure, the library can add support for arbitrarily complex expressions but say tomorrow, WGSL adds a new function. The library now needs to be updated to support this new function. A site like ShaderToy, would ideally not have write any new code for that function to be available to its users. If WebGPU provided the reflection info then the correct values would be computed for any new features provided by the implementation. With the current situation, any site that needs reflection to support user supplied shaders will need to be constantly updated to support new WGSL features.

**Comment by greggman** [comment]
I want to add, recently I wanted to write a wrapper. To do it correctly I needed to know BindGroup layouts so that I know at draw time, which bindgroups are actually being used. In other words, the wrapper tracks calls to `setBindGroup` and then at draw time it wants to check some stuff but it does not know which bindGroups are actually being used. Bindgroups might be assigned to indices 0, 1, 2, 3 but the pipeline might only use indices 0 and 2 but the wrapper can't know this without including an entire WGSL parser.

**Comment by eliemichel** [comment]
This is a nice suggestion. 👌 

However, contrary to what the original post suggests when noting that *"all the native APIs provide some sort of shader reflection information"*, I believe that acess to reflection should only be a **transpile-time** operation, rather than something that can be queried at any time. The host application can then any information they would like to retain.

This would make implementing such a feature much easier (it is now "just" about exposing bits of tint/naga that are already around anyways), without relying on any reflection capacity provided by underlying graphics APIs (making it more flexible for WebGPU to decide what to expose).

It could be exposed as an extension of the `ShaderModuleDescriptor`, which provides a callback that receives a shader introspection suite of function, valid only for the duration of the callback in question.

**Comment by 7ombie** [comment]
I was thinking of extending [the `createShaderModule` method][mdn] (on the device instance). That way, the source only needs to be parsed once, and the information is immediately returned (and retained by the caller until they're done with it).

The `createShaderModule` method takes a hash of arguments (`code`, `hints`, `label` and `sourceMap`). We could add two optional keys, named something like `compile` (which defaults to `true`) and `analyze` (which defaults to `false`), so we have the option of just compiling, just analyzing the source or doing both, and the method stays backwards compatible.

The actual information (when requested) would be directly bound to one or more new properties of the returned `GPUShaderModule` instance.

[mdn]: https://developer.mozilla.org/en-US/docs/Web/API/GPUDevice/createShaderModule

**Comment by eliemichel** [comment]
> For my part, I was thinking of extending the [createShaderModule](https://developer.mozilla.org/en-US/docs/Web/API/GPUDevice/createShaderModule) method (on the device instance).

That's what I meant as well, although I am not sure whether `analysis` would be the same thing for every use case, hence the possibility to either give more details in the request/descriptor than `analyze: true`, or that `analysis` is a object that we get once, can query and then release even if we keep on using the module (we don't want the module to have to remember about this information for its entire lifetime

> It seems unlikely this info will be available synchronously. The whole API of WebGPU is async. When you get a module back from device.createShaderModule that doesn't mean the module was actually created or compiled.

Faire point, a typicall use case would be to use this reflection information to build the pipeline though, so how could we work this around? Should we have an indirection between the **parsed** shader module and the **compiled** shader module? Or have `createShaderModule` return a pair of a module and a promise to the reflection object, the former of which could be explicitly awaited?

**Comment by greggman** [comment]
As I mentioned above, there is already [`getCompliationInfo`](https://gpuweb.github.io/gpuweb/#dom-gpushadermodule-getcompilationinfo). It seems like a natural place to get this info. For now though I'd suggest you use [wgsl_reflect](https://github.com/brendan-duncan/wgsl_reflect) and prototype.

For example

```
{
  const s_moduleToSource = new WeakMap();

  GPUDevice.prototype.createShaderModule = (function(origFn) {
    return function(desc) {
      const module = origFn.call(this, desc);
      if (module) {
        s_moduleToSource.set(module, desc.code);
      }
      return module;
    };
  })(GPUDevice.prototype.createShaderModule);

  GPUShaderModule.prototype.getCompilationInfo = (function(origFn) {
    return async function() {
      const info = await origFn.call(this);
      await applyReflectionInfo(this, info);
      return info;
    };
  })(GPUShaderModule.prototype.getCompilationInfo);

  async function applyReflectionInfo(module, info) {
    // this will only be loaded once, on demand.
    const { WgslReflect } = await import("https://brendan-duncan.github.io/wgsl_reflect/wgsl_reflect.module.js"); 
    const code = s_moduleToSource.get(module);
    const reflect = new WgslReflect(code);
    // do whatever transformation to "relfect" you think should be the standard
    info.reflect = reflect;
    return info;
  };
}
```

Now you can get the reflect info like this


```
const module = device.createShaderModule({code});
const info = await module.getCompliationInfo();
console.log(info.reflect);
```

[working example](https://jsgist.org/?src=27d1b5b3bc0190259a65f7c3df59c912)

Now you can try to figure out what works for you. What you think best format for the reflection data, etc...

**Comment by greggman** [comment]
One other issue in standardizing this; It's possible that any description of a "struct in memory" will be more difficult to standardize as there may be other APIs that also want a similar definition.

In my ideal world, you could do something like this

```
const module = device.createShaderModule({code});
const info = await module.getCompliationInfo();
const structDef = info.structs.Material;
const ab = new ArrayBuffer(structDef.size);
// Make a view of a `ab`
const structView = new StructView(ab, structDef);
```

Let's assume the `Material` is

```
struct Light {
  position: vec3f,
  color: vec3f,
  power: f32,
};

struct Material {
  lights: array<Light, 2>;
  color: vec3f,
  shininess: f32,
  roughness: f32,
};
```

Then I'd like to be able to do this

```
// Make a view of a `ab`
const structView = new StructView(ab, structDef);
struct.lights[1].position = [1, 2, 3];
struct.roughness = 456.
struct.set({
  shininess: 1.23,
  color: [0.5, 0.7, 1],
});
```

BUT, the issue is this is useful outside of WebGPU. For example, Wasm would benefit to be able to use this as well. So, a much larger community will want a say in what `structDef` is and `StructView`

note that [webgpu-utils](https://github.com/greggman/webgpu-utils) already provides much of this.

**Comment by eliemichel** [comment]
> there is already [getCompliationInfo](https://gpuweb.github.io/gpuweb/#dom-gpushadermodule-getcompilationinfo). It seems like a natural place to get this info

Your example makes it nice to use, my concern was that this reflection info could get big so we may want a way for the user to say "I'm done using reflection for this module, release it entierly from memory".

> So, a much larger community will want a say in what structDef is and StructView

I believe we can have `info` return basic size/offset info, then leave `StructView` up to third party libraries, there is no need for built-in support in the browser (one can use [DataView](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DataView) for instance).

**Comment by kainino0x** [comment]
StructView can be done in userspace, but it probably can't be both nice to use and efficient. Example that is nice but not efficient: https://github.com/kainino0x/structured-accessor

**Comment by eliemichel** [comment]
In this kind of situation I sometimes use the [`Function`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/Function) constructor to "pre-interpret" (if that makes sense) the view constructor... I know, that's almost an `eval` crime, but in other cases I could wee great performance boost! (And it's must safer than `eval`.)

Here is an example, with a less fancy API than yours but that gives a proof of concept which builds the view 30x faster (from what I could test): https://gist.github.com/eliemichel/56d8013093b881c39e9b5de72f63e7e5

**Comment by kainino0x** [comment]
Ahh, I should never say something can't be done in JS :) That's cool!

**Comment by 7ombie** [comment]
There's a (Stage 2) [JS struct proposal][js-struct], and the need to support WebGPU has been raised there. Not sure if it'll move forward.

[js-struct]: https://github.com/tc39/proposal-structs

**Comment by greggman** [comment]
AFAICT the JS struct proposal has nothing to do with memory layout so it doesn't seem relevant here.

**Comment by greggman** [comment]
I'm not so sure building strings and evaling them via `Function` has any perf benefits. Here's 2 implementations without `Function`. At least in Chrome and Safari they run around the same speed +/-

* [version 1](https://gist.github.com/greggman/a087a43f08f1f02c1934727dcbde971a)
* [version 2](https://gist.github.com/greggman/dda4766f7a9507f922f2abaa1063fb3b)

Here's [the first one I wrote](https://gist.github.com/greggman/ca3542a1f626c8786f8eb430fdaa20b7). It's slower, at least in Chrome, mostly because it uses `map` and `entries` which are known to be slower in v8.

Running them 500 times (10000 iterations each) and averaging the times I get

| browser | v1 | v2 | version from [comment above](https://github.com/gpuweb/gpuweb/issues/624#issuecomment-2451034143) |
| -- | -- | -- | -- |
| chrome | 6.213999999463558 ms | 6.044000000059604 ms | 6.206000000238419 ms |
| safari | 2.188 ms | 2.188000000000002 ms | 2.046 ms |

**Comment by kainino0x** [comment]
IIRC the JS struct proposal would help WebGPU with multithreading/SharedObjectTable (#747), and while they could relate to fixed-layout struct views onto ArrayBuffer/SharedArrayBuffer, that isn't the proposal today.

**Comment by eliemichel** [comment]
> note: there's a bug in the code linked in https://github.com/gpuweb/gpuweb/issues/624#issuecomment-2451034143 in that it's not subtracting begin time from end.

Indeed, fixed.

> I'm not so sure building strings and evaling them via Function has any perf benefits.

Mmh, I added a Function-less baseline to [the gist](https://gist.github.com/eliemichel/56d8013093b881c39e9b5de72f63e7e5) from above and I still see a noticeable difference on my end (on Windows, still 500 runs of 10000 iterations each):

| browser | Using `Function` | "Naive" baseline |
| ------------- | ------------- | ------------- |
| chrome | **11.89 ms** ± 1.43 ms  | 15.46 ms ± 1.55 ms |
| firefox | **19.58 ms** ± 4.95 ms  | 29.01 ms ± 5.81 ms |

> Also, I'm not sure that Function is safer than eval. This executes code

I meant safer against non-adversarial programmer mistakes (because variables inside the `Function` are better scoped) but it is in no case safe to run arbitrary user-provided code, indeed, I should have precised

**Comment by greggman** [comment]
As pointed out above, you still have to optimize the stringless version. Using `for (const [k, v] of Object.entries()}` is slow (At least on v8) compared to the optimized versions linked above.

**Comment by greggman** [comment]
Visiting this again

I'm not totally convinced devs actually want this unless it's synchronous. I think, like anything "free", devs will say want it, but, if we provide it async, they'll try it, realize they hate refactoring to be async, and end up not using it.

As it is now, AFAIK, most of the libraries out there are synchronous when it comes to setup. The only parts that aren't are resource loading.

I'm not sure how to test my theory. But, if I'm right and the committee won't allow sync reflection and that devs won't actually use async reflection, then it'd be a waste of time to add it as they'll just continue to use [wgsl_reflect](https://github.com/brendan-duncan/wgsl_reflect) or [related libraries](https://github.com/greggman/webgpu-utils) which are sync.

TL:DR; before we implement reflection we should decide if we can make it synchronous. If not, we should evaluate if devs will use it if it's async.

**Comment by eliemichel** [comment]
For the record, we tried a couple of weeks ago to draft what this reflection API could look like. We wanted to write a PoC and cleanup the writeup before sharing, but it may already fuel the discussion: [wgpuextensionrequest.pdf](https://github.com/user-attachments/files/20966601/wgpuextensionrequest.pdf)

**Comment by greggman** [comment]
I would suggest you take a look at [wgsl_reflect](https://github.com/brendan-duncan/wgsl_reflect) and it's usage in [webgpu-utils](https://github.com/greggman/webgpu-utils).

You can get struct info but to get bind group layouts you need entry points. That's because this is valid WGSL

```wgsl
@group(0) @binding(0) var t1: texture_2d<f32>;
@group(0) @binding(0) var<uniform> t2: vec4f;  // same binding point!

@compute @workgroup_size(1) fn cs1() {
  _ = t1;
}

@compute @workgroup_size(1) fn cs2() {
  _ = t2;
}
```

[parsed](https://webgpufundamentals.org/webgpu/lessons/resources/wgsl-offset-computer.html#x=5d000001006101000000000000003d888b0237284c928ba3d48de3c519adb7a1ce614b18d749198df308838de9489329e1fa722adfdacc29c0ce794394fd429be41d2f884202df187ea7c1ff5ca180bcfbc646f3daca51b1ba9902847f53912774fd62732c2caddf9c6554bb2049f7fe8c5fee9567c865764f5ab62633f7f6d411d85d7e1aacc57deedf6ed57adda099237b0c42e88694c066c8e57522417308ef25b4532d8ac304fe611f96d576cafa12bac28c218dfc9f3e82e7e700b94d09ba83521a6ac43078ae786bdba55fc0812039eca74c30b479f5f02021d535584f77fafa9b689eff0792dc00)

To make it clear, you also can't just return premade bind group layouts since you don't know the layout until the user tells you the pairings.

```wgsl
@group(0) @binding(0) var tv1: texture_2d<f32>;
@group(0) @binding(0) var<uniform> tv2: vec4f; // same binding as previous
@group(0) @binding(1) var tf1: texture_2d<f32>;
@group(0) @binding(1) var<uniform> tf2: vec4f; // same binding as previous

@vertex fn vs1() -> @builtin(position) vec4f { _ = tv1; return vec4f(0); }
@vertex fn vs2() -> @builtin(position) vec4f { _ = tv2; return vec4f(0); }

@fragment fn fs1() -> @location(0) vec4f { _ = tf1; return vec4f(0); }
@fragment fn fs2() -> @location(0) vec4f { _ = tf2; return vec4f(0); }
```

[parsed](https://webgpufundamentals.org/webgpu/lessons/resources/wgsl-offset-computer.html#x=5d00000100a002000000000000003d888b0237284c928ba3d48de3c519adb7a1ce614b18d749198df308838de948964a919dcbf92919be165a4e58bb56acf847015b056bd4c9afbd99004c6e593bba8e09f3fcacb006521e1f505f5b972961a67b05daf75490608a24539aae4e19b6ca543402efd42adbe874e904b7e9fba8c1741fce5a3efd46d2bfd2c9dbb87373bfe8cc3633c9812f3d5a6206d2f72b357cd94917db0c36a2ee1f42d0e91dd96f1898a48f6d3fc2b62a84aaa76c70d0c1f17cb1407e65f44cbf7503b069738fb4c758b23db8fb656f9d842005d556c6442464eceddef6c5f842a664ccbc02b279f42703dd48d0590043ffd231c9a90db858daaac4f4aac70f889635eedd4d7f54d8bc8068a0df876798082c2c88678c8cdffea92079)

The module above can create 4 pairs (vs1+fs1), (vs1+fs2), (vs2+fs1), (vs2+fs2). And of course, when you pair them they could be in separate modules.

[webgpu-utils](https://github.com/greggman/webgpu-utils) basically has a function to give you an array of bindgroup layout descriptors. It takes as input a subset of a `GPURenderPipelineDescriptor` / `GPUComputePipelineDescriptor` so that you can re-use the pipeline descriptor you were going to create anyway with multiple modules and default or specified entry points. To handle the defaults means you need a list of typed entry points (a list of entry points and knowing if they are vertex,fragment,compute) and each of those entry points needs to list its resource bindings.

Given a subset of pipleine descriptor, [that library returns an array of bindgrouplayout descriptors](https://greggman.github.io/webgpu-utils/docs/functions/makeBindGroupLayoutDescriptors.html) such that you can create the bindgrouplayouts and pipelinelayout. To be clear, I'm not suggesting your API provide this functionality. I'm just giving a working example of a usecase and the data needed to support that usecase.
